EIPS/eip-7867.md

```diff
@@ -0,0 +1,553 @@
+        "value": "0x182183",
+        "data": "0xfbadbaf01",
+        "capabilities": {
+            "flow-control": {
+                "on-failure": "continue"
+            }
+        }
+      }
+    ]
+  }
+]
+```
+
+#### `wallet_getCapabilities`
+
+The following JSON Schema is inserted into the per-chain object returned from `wallet_getCapabilities` with a key of `flow-control`.
+
+##### Schema
+
+```json
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "type": "object",
+    "additionalProperties": false,
+    "minProperties": 1,
+    "properties": {
+        "none": { "$ref": "#/$defs/on-failure" },
+        "loose": { "$ref": "#/$defs/on-failure" },
+        "strict": { "$ref": "#/$defs/on-failure" }
+    },
+    "$defs": {
+        "on-failure": {
+            "type": "array",
+            "uniqueItems": true,
+            "minItems": 1,
+            "items": {
+                "enum": ["rollback", "halt", "continue"]
+            }
+        }
+    }
+}
+```
+
+##### Example Response
+
+```js
+{
+    "0x1": {
+        "flow-control": {
+            "loose": ["halt", "continue"],
+            "strict": ["continue"]
+        }
+    }
+}
+```
+
+### Concepts
+
+#### Call Failure
+
+<!-- TODO: is a "call failure" the same as a "revert"? -->
+
+#### Rollback
+
+A rollback is informally defined as "causing no meaningful changes on chain." A
+rolled back batch only makes gas accounting and bookkeeping (eg. nonce)
+changes. In other words, a rollback is the default behaviour of EIP-5792 when
+a call fails.
+
+#### Critical Calls
+
+A critical call is a call that causes the entire batch to rollback on failure,
+and correspondingly a non-critical call does not. Specifically, a critical call
+has a call-scope `on-failure` of `rollback` (or no `on-failure` present),
+while non-critical calls have either `halt` or `continue`.
+
+#### Atomicity Levels
+
+This proposal introduces three atomicity levels: strict, loose, and none; enabled
+by setting batch-scope `atomicity` to `strict`, `loose`, or `none` respectively.
+Strict may also be enabled by omitting `atomicity` entirely.
+
+Strict atomicity is simply naming the default behaviour of EIP-5792: calls
+within a single batch MUST be contiguous and applied atomically (or the batch
+rolled back.)
+
+Loose atomicity, on the other hand, is a weaker guarantee. In the event of a
+block reorg, any number of calls from the batch MAY appear on chain (possibly
+interspersed with other transactions). If there are no block reorgs, loose
+atomicity MUST provide the same guarantees as strict.
+
+The none level of atomicity only provides the guarantee that the calls appear on
+chain in the order they are in the batch. Any number of calls from the batch
+MAY appear on chain (possibly interspersed with other transactions).
+
+### Behaviour
+
+#### `wallet_sendCalls`
+
+The wallet MUST reject `wallet_sendCalls` requests with error code 0
+<!-- TODO --> where both:
```

Comment:
As mentioned above, this creates a bad UX on mobile devices. We should rather default to one agreed behaviour to keep the user flow consistent instead of letting a wallet auto-reject such requests:
https://github.com/ethereum/EIPs/pull/9259/files#r1948151108